prime-dotnet-webapi/Services/EnrolleePaperSubmissionService.cs

```diff
@@ -264,5 +264,126 @@ public async Task FinalizeSubmissionAsync(int enrolleeId)
+        {
+            var confirmedLinks = await _context.EnrolleeLinkedEnrolment
+                .AsNoTracking()
+                .Where(cl => cl.Confirmed)
+                .Select(cl => cl.PaperEnrolleeId)
+                .ToListAsync();
+
+            return await _context.Enrollees
+                .AsNoTracking()
+                .AnyAsync(
+                    e => e.GPID.StartsWith(PaperGpidPrefix)
+                    && !confirmedLinks.Contains(e.Id)
+                    && e.DateOfBirth.Date == dateOfBirth.Date
+                );
+        }
+
+        public async Task<IEnumerable<Enrollee>> GetPotentialPaperEnrolleeReturneesAsync(DateTime dateOfBirth)
+        {
+            var confirmedLinks = await _context.EnrolleeLinkedEnrolment
+                .AsNoTracking()
+                .Where(cl => cl.Confirmed)
+                .Select(cl => cl.PaperEnrolleeId)
+                .ToListAsync();
+
+            return await _context.Enrollees
+                .AsNoTracking()
+                .Where(
+                    e => e.GPID.StartsWith(PaperGpidPrefix)
+                    && e.DateOfBirth.Date == dateOfBirth.Date
+                    && !confirmedLinks.Contains(e.Id)
+                )
+                .ToListAsync();
+        }
+
+        public async Task<bool> LinkEnrolmentToPaperEnrolmentAsync(int enrolmentId, int paperEnrolmentId, bool isConfirmed = false)
+        {
+            var enrollee = await _context.Enrollees
+                .Where(
+                    e => e.Id == enrolmentId
+                    && _context.EnrolleeLinkedEnrolment.Any(link => link.EnrolleeId == e.Id)
+                )
+                .AnyAsync();
+
+            var paperEnrollee = await _context.Enrollees
+                .Where(
+                    pe => pe.GPID.StartsWith(PaperGpidPrefix)
+                    && pe.Id == paperEnrolmentId
+                    && _context.EnrolleeLinkedEnrolment.Any(
+                        link => link.PaperEnrolleeId == pe.Id
+                        )
+                )
+                .AnyAsync();
+
+            if (paperEnrollee)
+            {
+                return false;
+            }
+
+            var enrolleeLinkedEnrolment = await _context.EnrolleeLinkedEnrolment
+                .Where(ele => ele.EnrolleeId == enrolmentId)
+                .SingleOrDefaultAsync();
+
+            enrolleeLinkedEnrolment.PaperEnrolleeId = paperEnrolmentId;
+            enrolleeLinkedEnrolment.EnrolmentLinkDate = DateTime.Now;
+            enrolleeLinkedEnrolment.Confirmed = isConfirmed;
+
+            await _context.SaveChangesAsync();
+
+            return true;
+        }
+
+        public async Task CreateInitialLinkAsync(int enrolleeId, string userProvidedGpid)
+        {
+            var enrolleeLinkedEnrolment = await _context.EnrolleeLinkedEnrolment
+                .Where(ele => ele.EnrolleeId == enrolleeId)
+                .SingleOrDefaultAsync();
```

Comment:
since filtering followed by materializing is such a common operation, `.Any()`, `.Single`, `.First()`, `.SingleOrDefault()`, and `.FirstOrDefault()` all have overloads that let you specify the predicate in one line. so this code can just be:
```c#
var enrolleeLinkedEnrolment = await _context.EnrolleeLinkedEnrolment
    .SingleOrDefaultAsync(ele => ele.EnrolleeId == enrolleeId);
```